### DISCUSSION TITLE
limit_strideの処理

### DISCUSSION THREAD

**Discussion by Naoki-Hiraoka** [YutaKojio/hrpsys-base#56] [Q&A] [answered]
現状ほとんど`gg.stride_limitation_type = OpenHRP.AutoBalancerService.CIRCLE`で運用されていますが、
https://github.com/YutaKojio/rtmros_tutorials/blob/ba5f3ec4cc8e5d8f4067a93ecf9ba29f3ca02c8c/hrpsys_ros_bridge_tutorials/src/hrpsys_ros_bridge_tutorials/urata_hrpsys_config.py#L350

以下の2点について、重要な理由がある意図的なものなのか、そうではないものなのか、教えていただけますでしょうか。

1. goPosやgoVelocityのfootstep生成時に使用される`gait_generator::limit_stride`では円を用いた着地位置制限を行っているが、
https://github.com/YutaKojio/hrpsys-base/blob/bacb4a0a2c099ae0e4fda7aa4f696fbf289e4111/rtc/AutoBalancer/GaitGenerator.cpp#L1452-L1459
着地位置修正時(vision無し)に使用される`gait_generator::limit_stride_rectangle `では行っていない点

2. goPosやgoVelocityのfootstep生成時に使用される`gait_generator::limit_stride`や、着地位置修正時(vision無し)に使用される`gait_generator::limit_stride_rectangle `では、遊脚のyaw旋回に応じてinsideの制約が変化するが、
 https://github.com/YutaKojio/hrpsys-base/blob/bacb4a0a2c099ae0e4fda7aa4f696fbf289e4111/rtc/AutoBalancer/GaitGenerator.cpp#L1466-L1474
https://github.com/YutaKojio/hrpsys-base/blob/bacb4a0a2c099ae0e4fda7aa4f696fbf289e4111/rtc/AutoBalancer/GaitGenerator.cpp#L1494-L1502
着地位置修正時(vision有り)に使用される`gait_generator::limit_stride_vision`では遊脚のyaw旋回を考慮しない点
https://github.com/YutaKojio/hrpsys-base/blob/bacb4a0a2c099ae0e4fda7aa4f696fbf289e4111/rtc/AutoBalancer/GaitGenerator.cpp#L831-L841

**Comment by YutaKojio** [comment, marked as answer]
どちらの指摘もあっています．
意図的というか，ちゃんとやっていないことはわかっていてやっていない状態です．

1. visionなしのIROS2019の段階では，着地位置・時間修正でx成分とy成分が分離できるようにしたくて長方形にしました．
しかし，visionありのIROS2020では，着地位置制限は凸包であればなんでもよくなったので，CIRCLE制限に近い凸包で近似したものを使ってもいい気はします．

2. visionありの，足場制限と着地位置制限の積集合への変換は一歩の最初に一回のみ行っていて，それ以降に変わるyawにちゃんと対応できないので，この時点でもyawを考慮しないことにしました．
ただ，yawの目標値が遊脚途中で変わることはあまりないので，この時点でも考慮するようにしてもいいと思います．
また，そもそもこの積集合への変換がどの程度の計算量になっているのか真面目に考えていないので，軽いようであれば毎制御周期行うようにして，最新の足場情報へ更新できるようにしてもいいし，重いようなら実装をもっと賢いものに変えるのもありだと思います．
